### PR TITLE
kiwix-serve displays both protocol attached ips

### LINKED ISSUE #703: [REGRESSION] kiwix-serve URL helper does not display iPv4 URL anymore
With cutting-edge dev kiwix-serve the URL help to join the server via HTTP is not anymore really what it should.

I get:
```
$ kiwix-serve --port=8081 /home/kelson/Downloads/*zim
The Kiwix server is running and can be accessed in the local network at: http://[::1]:8081
```

My system has both IPv4 AND IPv6, so both should be displayed in the helper. It should look like (if -i=all or default):
```
$ kiwix-serve --port=8081 /home/kelson/Downloads/*zim
The Kiwix server is running and can be accessed in the local network at: http://127.0.0.1:8081 or http://[::1]:8081
```

### PR DESCRIPTION
**Pull request by sgourdas**
This changes the logic to use the refactored `getBestPublicIps` so kiwix-serve can display all attached ips based on the protocol that was requested by the user.

Fixes #703

Relies on https://github.com/kiwix/libkiwix/pull/1132

### CODE REVIEW

**Review by kelson42 — APPROVED** [review]
LGTM from a user perspective. Ready for code review.

**Comment by veloman-yunkan** [comment]
> Ready for code review.

Not yet. Comments from my previous review have not been addressed.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

**Comment by veloman-yunkan** [comment]
> > Ready for code review.
> 
> Not yet. Comments from my previous review have not been addressed.

Oops! It turns out that I failed to publish my review - it was in pending state since last Wednesday :(

File: src/server/kiwix-serve.cpp
Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: Essentially the logic remains the same as before. How it is setup is that the `InternalServer::start` is going to set the `m_ipMode` as ipv6 (previously all) if a valid v6 ip has been provided by the user.

File: src/server/kiwix-serve.cpp
Comment (sgourdas) [reply src/server/kiwix-serve.cpp:392]: I strongly think that we should trust what the server sets for the final print message. This an additional reason for line 381 here, because some mode handling is being done in the server as well, as mentioned in my other comment.

What do you think?

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:329]: Maybe. But codewise it appears flawed. I think that the interaction of `Server::setIpMode()` and `Server::setAddress()` should be defined in such a way that the former takes precedence. In other words, if an IPv4 mode is requested then only IPv4 addresses should be accepted.

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: So essentially we should:
- Remove all server mode handling and move it to kiwix-serve (removing also [kiwix-serve.cpp:381](https://github.com/kiwix/kiwix-tools/pull/704/files#diff-29eedb326cc54203c610d3910176face715ff30664857e7d8ced287ff455551aR381)).
- Set default mode (no mode through `-i` provided) to `all` (also, [server.h:84](https://github.com/kiwix/libkiwix/blob/0b14fda94d6d7047bbcc9b00cb53f261fcd66880/include/server.h#L84) should be changed).

Correct?

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:392]: I think that extra complexity only hurts. Here we only want to print the server addresses in use. Those addresses come from `Server::getAddress()` and we shouldn't rely on anything else to print them.

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:329]: > * Remove all server mode handling and move it to kiwix-serve

I don't know what is covered by this so cannot answer

> * Set default mode (no mode through `-i` provided) to `all`

It will work, but IMHO it's better to add an extra `enum` member `IpMode::auto`

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: Would `IpMode::auto` have any use case other than `IpMode::all`'s?

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: Or just an easier way for kiwix-serve to understand that no ip mode has been supplied?

Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:329]: > Would `IpMode::auto` have any use case other than `IpMode::all`'s?

It was included in my [original proposal](https://github.com/kiwix/libkiwix/pull/1074#pullrequestreview-2019253469) for enhancing the `kiwix-serve` CLI with support for IPv6, but did not make way into implementation. Whatever the reason for that (you can find it in the ensuing discussion) it has a comeback at least in the C++ code.

> Or just an easier way for kiwix-serve to understand that no ip mode has been supplied?

Mostly yes, but with some potential to find other uses later.

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: Alright, I see how it can be useful in future scenarios. I will add it and make it the default mode for when an address is provided or no -i at all.

P.S. `auto` is not allowed as a name. Will use `flex`.

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:329]: To be honest it is a bit complicated with the current implementation to really put it in use, but I will wait for your next review after committing to discuss further.

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:335]: Why don't you handle `IpMode::flex` in `libkiwix`?

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:393]: I don't understand why you need these. I think that `Server::getAddress()` must return only the addresses actually in use by the server. If it doesn't work that way - fix it in `libkiwix`.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

**Comment by kelson42** [comment]
@sgourdas Where is the new `flex` option listed in the usage()?
```
 -i=<address> --address=<address>        Listen only on the specified IP address. Specify 'ipv4', 'ipv6' or 'all' to listen on all IPv4, IPv6 or both types of addresses, respectively [default: all]
```

**Comment by sgourdas** [comment]
@kelson42 the new `FLEX` enum is an **internal** use only mode helping to signify, currently, a state where the server should choose the protocol. It is also a good future proof way to help pass information back and forth with the server.

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:330]: Why do we still have `FLEX`? The purpose of capitalizing the `IpMode` enum members was so that we could use the more intuitive word `AUTO`.

Comment (sgourdas) [reply src/server/kiwix-serve.cpp:330]: Sorry for this. Missed this point after coming back. Will fix on next iteration.

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:332]: Long lines are not good.

Reformat as a multiline statement with braces (like it was in the previous iteration).

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
This is mostly ready to merge, but since it depends on a PR in `libkiwix`, I am playing picky and deferring the approval until the libkiwix PR is merged.

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:335]: 1. You lost the declaration of `ipMode`.
2. Please minimize the diff so that that it's obvious that this change is merely about the capitalization of the `IpMode` enum members.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/server/kiwix-serve.cpp
Comment (veloman-yunkan) [inline src/server/kiwix-serve.cpp:335]: 1. You lost the declaration of `ipMode`.
2. Please minimize the diff so that that it's obvious that this change is merely about the capitalization of the `IpMode` enum members.

Comment (veloman-yunkan) [reply src/server/kiwix-serve.cpp:335]: > 2. Please minimize the diff so that that it's obvious that this change is merely about the capitalization of the IpMode enum members.

... and the default IP mode becoming `AUTO`.

**Comment by sgourdas** [comment]
@kelson42 missed a line. Try again please.

**Review by veloman-yunkan — APPROVED** [review]

**Comment by kelson42** [comment]
LGTM

**Pull request merged**
